pkg/metallb/frr.go

```diff
@@ -0,0 +1,589 @@
+
+	return builder
+}
+
+// WithToReceiveModeFiltered allows only specified prefixes to be received.
+func (builder *FrrConfigurationBuilder) WithToReceiveModeFiltered(prefixes []string,
+	routerIndex, neighborIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Allows the FRR nodes to receive specific routes in namespace %s in mode type: filtered",
+		builder.Definition.Namespace)
+
+	// Check if the routerIndex is within bounds
+	if routerIndex < 0 || routerIndex >= len(builder.Definition.Spec.BGP.Routers) {
+		glog.V(100).Infof("Invalid routerIndex: %d", routerIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid routerIndex: %d", routerIndex)
+	}
+
+	// Check if the neighborIndex is within bounds
+	if neighborIndex < 0 || neighborIndex >= len(builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors) {
+		glog.V(100).Infof("Invalid neighborIndex: %d", neighborIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid neighborIndex: %d", neighborIndex)
+	}
+
+	// Validate CIDR prefixes
+	for _, prefix := range prefixes {
+		if _, _, err := net.ParseCIDR(prefix); err != nil {
+			glog.V(100).Infof("the frrConfiguration prefix %s is not a valid CIDR", prefix)
+			builder.errorMsg = fmt.Sprintf("the prefix %s is not a valid CIDR", prefix)
+		}
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
+
+	// Prepare the list of allowed prefixes
+	allowedPrefixes := make([]frrtypes.PrefixSelector, len(prefixes))
+
+	for i, prefix := range prefixes {
+		glog.V(100).Infof("AllowMode 'filtered' receives only routes within the prefix %v", prefix)
+		allowedPrefixes[i] = frrtypes.PrefixSelector{Prefix: prefix}
+	}
+
+	// Update the specific neighbor's ToReceive field
+	builder.Definition.Spec.BGP.Routers[routerIndex].Neighbors[neighborIndex].ToReceive.Allowed.Prefixes = allowedPrefixes
+
+	return builder
+}
+
+// WithBGPNeighbor defines a single neighbor IP and ASN number.
+func (builder *FrrConfigurationBuilder) WithBGPNeighbor(bgpPeerIP string,
+	remoteAS uint32, routerIndex int) *FrrConfigurationBuilder {
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	glog.V(100).Infof(
+		"Define BGP Neighbor %s with peer IP %s and ASN %v",
+		builder.Definition.Name, bgpPeerIP, remoteAS)
+
+	if net.ParseIP(bgpPeerIP) == nil {
+		glog.V(100).Infof("The peerIP for the frrConfiguration bgp peer contains invalid ip address %s",
+			bgpPeerIP)
+
+		builder.errorMsg = "frrConfiguration 'peerIP' of the BGPPeer contains invalid ip address"
+	}
+
+	// Check if the routerIndex is within bounds
+	if routerIndex < 0 || routerIndex >= len(builder.Definition.Spec.BGP.Routers) {
+		glog.V(100).Infof("Invalid routerIndex: %d", routerIndex)
+
+		builder.errorMsg = fmt.Sprintf("invalid routerIndex: %d", routerIndex)
+	}
+
+	if builder.errorMsg != "" {
+		return builder
+	}
```

Comment:
This can be removed if we short-circuit return builder.

Reply:
removed